src/main/java/io/vertx/core/http/impl/HttpClientRequestImpl.java

```diff
@@ -351,6 +378,42 @@ synchronized void handleResponse(HttpClientResponseImpl resp) {
     }
   }
 
+  private void performRedirect(HttpClientResponseImpl resp) throws URISyntaxException {
+    if (++redirectCount >= client.getOptions().getMaxRedirects()) {
+      throw new VertxException("Too many redirects: " + redirectCount);
+    }
+    String location = resp.getHeader(io.vertx.core.http.HttpHeaders.LOCATION);
+    URI uri = new URI(location);
+    URI absoluteUri;
+    if (!uri.isAbsolute()) {
+      absoluteUri = new URI(this.uri()).resolve(location);
+    } else {
+      absoluteUri = uri;
+      host = absoluteUri.getHost();
+      port = absoluteUri.getPort();
+    }
+
+    resp.endHandler((nothing) -> {
+      String newUri = absoluteUri.toString();
+      reset(newUri);
+      end();
+    });
+  }
+
+  private void reset(String newUri) {
```

Comment:
Ok, but then all access to the field must be guarded. It's not the case. So there are race conditions.